src/relay/ranker/mod.rs

```diff
@@ -0,0 +1,167 @@
+                break;
+            }
+            count += 1;
+        }
+        self.failure_times.len() - count
+    }
+
+    /// Return the mean failure rate per hour based on the known failures happened during the specified recent period.
+    fn mean_failure_rate(&self, period: Duration) -> f64 {
+        let failures = self.failures_last(period);
+
+        if failures == 0 {
+            return 0.0;
+        }
+        failures as f64 / (period.as_secs_f64() / 3600.0)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RelayRanker {
+    relay_stats: Arc<Mutex<RefCell<HashMap<String, RelayStats>>>>,
+    max_duration: Duration,
+}
+
+impl RelayRanker {
+    pub fn new(retain: Duration) -> RelayRanker {
+        RelayRanker {
+            relay_stats: Arc::new(Mutex::new(RefCell::new(HashMap::new()))),
+            max_duration: retain,
+        }
+    }
+
+    /// report a service failure to the ranker
+    pub fn downvote(&self, domain: impl Into<String>) -> Result<()> {
+        let guard = match self.relay_stats.lock() {
+            Ok(guard) => guard,
+            Err(_) => return Err(anyhow::anyhow!("failed to acquire lock")),
+        };
+        let mut inner = guard.borrow_mut();
+        let stats = inner.entry(domain.into()).or_insert(RelayStats::new());
+        stats.add_failure(self.max_duration);
+        Ok(())
+    }
+
+    /// Sort the domains of relays in ascending order based on their recent failure rate.
+    ///
+    /// The ranking of relays is determined by the number of failures that occur during a specified period of time.
+    /// This ensures that the affected relay’s rank will improve over time, and messages will be routed to it again if the service recovers.
+    /// If multiple relays have the same failure rate, their order will be randomized
+    pub fn reorder(&self, domains: &mut Vec<&str>) -> Result<()> {
+        let guard = match self.relay_stats.lock() {
+            Ok(guard) => guard,
+            Err(_) => return Err(anyhow::anyhow!("failed to acquire lock")),
+        };
+
+        domains.sort_by(|a, b| {
+            let a_failure_rate = self._mean_failure_rate(*a, &guard);
+            let b_failure_rate = self._mean_failure_rate(*b, &guard);
+            if a_failure_rate == b_failure_rate {
+                let mut rng = rand::thread_rng();
+                rng.gen::<bool>().cmp(&rng.gen::<bool>())
+            } else {
+                a_failure_rate.total_cmp(&b_failure_rate)
+            }
+        });
+        log::debug!("ranking system hint: {:?}", domains);
+        Ok(())
+    }
+
+    fn _mean_failure_rate(
+        &self,
+        domain: impl Into<String>,
+        guard: &MutexGuard<'_, RefCell<HashMap<String, RelayStats>>>,
+    ) -> f64 {
+        let inner = guard.borrow();
+        if let Some(stats) = inner.get(&domain.into()) {
+            stats.mean_failure_rate(self.max_duration)
+        } else {
+            0.0
+        }
+    }
+}
```

Comment:
you kinda made this a little bit complex (the argument list) 

when u have the lock of the map, then you get the guard which deref to the actual inner type. so basically the code can be as simple as 
```rust
let mut map = self.realy_stats.lock().await;
// then u do simply 
let mut stat = map.get_mut("name");
// NOTE: the get_mut is a method of hashmap not guard
// then u can pass stat to other method or use it directly!
// no need to pass the guard ever as a guard
```

Reply:
thanks done